src/take_while_inclusive.rs

```diff
@@ -0,0 +1,68 @@
+use core::iter::FusedIterator;
+use std::fmt;
+
+/// An iterator adaptor that consumes elements while the given predicate is
+/// `true`, including the element for which the predicate first returned
+/// `false`.
+///
+/// See [`.take_while_inclusive()`](crate::Itertools::take_while_inclusive)
+/// for more information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct TakeWhileInclusive<'a, I: 'a, F> {
+    iter: &'a mut I,
+    predicate: F,
+    done: bool,
+}
+
+impl<'a, I, F> TakeWhileInclusive<'a, I, F>
+where
+    I: Iterator,
+    F: FnMut(&I::Item) -> bool,
+{
+    /// Create a new [`TakeWhileInclusive`] from an iterator and a predicate.
+    pub fn new(iter: &'a mut I, predicate: F) -> Self {
+        Self { iter, predicate, done: false}
+    }
+}
+
+impl<'a, I, F> fmt::Debug for TakeWhileInclusive<'a, I, F>
+    where I: Iterator + fmt::Debug,
+{
+    debug_fmt_fields!(TakeWhileInclusive, iter);
+}
+
+impl<'a, I, F> Iterator for TakeWhileInclusive<'a, I, F>
+where
+    I: Iterator,
+    F: FnMut(&I::Item) -> bool
+{
+    type Item = I::Item;
```

Comment:
pondering: Would it be worth returning the result of the predicate in some way in the item type, since that information is known free anyway?

I was reminded of [`Position::Last`](https://docs.rs/itertools/latest/itertools/enum.Position.html#variant.Last), though it shouldn't be that type (since this doesn't know the other variants inherently), and I suppose one could always `.take_while_inclusive(…).with_position()` if they wanted.

My brain then thought about things like `Result<I::Item, I::Item>` so the last item is an `Err`.  But that makes me go one step further and think of a `take_while_map_inclusive` (hopefully with a better name!) where the predicate returns a `Result` (or a `ControlFlow`) instead of a `bool`, and stops after returning the `Err` (or the `Break`).

Dunno if any of that is useful or worth doing, but I figured I'd toss it out for discussion.

Reply:
Returning a custom type that Deref's to the actual item type but has additional information about the predicate could be a good idea.

Otherwise, the call site might be able to infer it:
All entries except the last one in the iterator had the predicate return true, and the last had the predicate return false, with the following exceptions:
- If the returned iterator is empty, then there is no last element that returned true.
-  If the returned iterator has the same length as the original iterator, then there is no last element that returned true

Reply:
Oh, I should have read the Iterator docs better.  It [looks](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map_while) like what I propose would be named `map_while_inclusive` to go with `take_while_inclusive`.

I'll add that to the ACP, if one gets opened.